aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -114,6 +114,46 @@ static std::unique_ptr<TensorIterator> make_reduction(
   return TensorIterator::reduce_op(viewed_result, self.to(dtype));
 }
 
+static std::unique_ptr<TensorIterator> make_reduction(
+    const char* name, Tensor& result1, Tensor& result2, const Tensor& self, c10::optional<IntArrayRef> dim,
+    bool keepdim, ScalarType dtype)
+{
+  // check that result type and dtype match if provided
+  AT_CHECK(
+      !result1.defined() || result1.type().scalarType() == dtype,
+      name, ": provided dtype must match dtype of result. Got ",
+      toString(result1.type().scalarType()),
+      " and ",
+      toString(dtype),
+      ".");
+
+  AT_CHECK(
+      !result2.defined() || result2.type().scalarType() == dtype,
+      name, ": provided dtype must match dtype of result. Got ",
+      toString(result2.type().scalarType()),
+      " and ",
+      toString(dtype),
+      ".");
```

Comment:
Let's avoid copying and pasting code. Two options here: move this check out into a separate function and call it for both result1 and result2, or do something like:

```c++
for (Tensor *t: {&result1, &result2}) {
  Tensor& result = *t;
  AT_CHECK(
      !result.defined() || result.type().scalarType() == dtype,
      name, ": provided dtype must match dtype of result. Got ",
      toString(result.type().scalarType()),
      " and ",
      toString(dtype),
      ".");
}
```

Reply:
Fixed